### ISSUE TITLE
Add notebook: open-set metric learning on CUB + CLIP zero-shot

### ISSUE THREAD

**Issue by yoavram** [yoavram/DataSciPy#11]
## Summary

Add `sessions/metric_learning.ipynb` to `sessions/`, taking CUB-200 (already used in
`transfer.ipynb`) into the open-set regime, and closing with CLIP zero-shot classification on
the same held-out classes.

The conceptual through-line is **proxies**. ArcFace class weights are learned proxies in a
normalized embedding space. CLIP text embeddings are proxies *generated from language*.
Zero-shot classification is nearest-proxy retrieval with no training. The CLIP section is the
payoff of the notebook, not an appendix — it must be framed that way, and it runs on the same
100 held-out CUB classes so the comparison to the supervised metric model is like-for-like.

This is one of two notebooks. The companion issue covers re-identification and deep local
descriptors on sea turtles.

`functional_keras.ipynb` (Siamese network on MNIST even/odd digits) is the toy-problem
introduction to embedding spaces and open-set generalization. This notebook is the real
version. **State that delta explicitly in the opening cell** — otherwise students read it as
"MNIST again, with birds". The delta is: proxy-based rather than pair-based losses, ranking
metrics rather than verification AUC, real data, and a pretrained backbone.

---

## Context and constraints

- Repo: `yoavram/DataSciPy`, branch `DL2026`.
- Stack: Keras 3.15.1, **JAX backend**. All custom layers and losses must use `keras.ops`
  only (no `tf.*`, no backend-specific calls).
- House style: notebooks ship with pre-trained `.keras` checkpoints; training cells present
  but **commented out**, with an active loading cell. Copy the pattern from `transfer.ipynb`.
- Course convention: images normalized to `[0, 1]`.
- Must execute end-to-end on CPU in a few minutes using shipped checkpoints and precomputed
  arrays. GPU work goes in standalone scripts.
- Backbone: **`keras.applications.EfficientNetV2S`**, ImageNet-1K weights.
- Placement in `index.ipynb`: immediately after `transfer.ipynb`.

---

## Known API traps — verify on Keras 3.15.1, then document in a markdown cell

Treat these the way the course already treats `RandomFlip`'s default mode and the
`value_range=(0, 255)` default on `RandomErasing`/`RandomContrast`.

1. **EfficientNetV2 preprocessing.** `keras.applications.EfficientNetV2S` defaults to
   `include_preprocessing=True` and expects `[0, 255]`; `preprocess_input` is a pass-through.
   The course normalizes to `[0, 1]`, so the default will silently mis-scale inputs. **Verify
   actual behaviour**, then pick one convention explicitly (pass `[0, 255]` and keep internal
   preprocessing, or set `include_preprocessing=False` and normalize to `[0, 1]`). Document
   the choice.
2. **ArcFace numerical instability.** `sin θ = sqrt(1 − cos²θ)` has infinite gradient at
   `cos θ → ±1`, producing NaNs. Clip `cos θ` to `[−1 + eps, 1 − eps]` before the sqrt.
   Name this as a gotcha; demonstrate it if cheap.
3. **BatchNorm on unfreeze.** Only relevant if the notebook fine-tunes rather than using
   frozen features (see "Compute" below). In Keras a BatchNormalization layer with
   `trainable=False` also runs in inference mode, and this interacts with the parent model's
   `trainable` flag. Standard cause of "fine-tuning made it worse".

---

### Setup

CUB-200-2011, reusing the loading code already in `transfer.ipynb`. Standard metric-learning
split (Song et al., 2016): classes 1–100 train, classes 101–200 test. **Test classes are never
seen during training.**

Opening narrative: closed-set classification assumes a fixed label set. Open-set
classification, few-shot learning, retrieval and re-identification all break that assumption.
The answer is to learn an embedding rather than a classifier.

**Framing note:** CUB classes are *species*, not individuals. The mathematics of open-set
retrieval is identical, but say so in one line so the re-identification motivation (picked up
in the companion notebook) isn't misleading.

### Contrastive learning — brief only

One or two markdown cells on pair/triplet losses and the mining problem (semi-hard negatives,
batch construction), pointing back to `functional_keras.ipynb`. Then the argument for moving
on: proxy-based losses compare each sample against all class proxies, avoiding the mining
machinery entirely. No triplet-loss implementation here.

### ArcFace

ArcFace does **not** require a two-input model. Implement it as:

- A custom layer (`ArcFaceHead`) holding weights of shape `(d, C)`, no bias. L2-normalize
  both the input embedding and the weight columns. Output `s · cos θ`.
- The angular margin applied **in the loss**, where `y_true` is already available. A custom
  loss recovers `cos θ` from the scaled logits, applies
  `cos(θ + m) = cos θ · cos m − sin θ · sin m` to the target class only, re-scales by `s`,
  and passes to categorical cross-entropy.
- Handle `θ + m > π` with the standard threshold trick (when `cos θ ≤ cos(π − m)`, fall back
  to `cos θ − m · sin(π − m)`), or the "easy margin" variant. Pick one and say which.
- Defaults: `s = 30`, `m = 0.5`, embedding dim 512. Expose `s` and `m` as notebook variables
  so students can sweep them — ArcFace is hyperparameter-brittle and that should be visible.
- At inference, discard the head: `keras.Model(inputs, embedding)`.

Write `ArcFaceHead` cleanly enough that the companion notebook can present its sub-center
variant as a modification exercise on this layer.

**Batch sampling.** Explain that proxy-based losses are *insensitive* to batch composition by
design — every sample sees every proxy, and ArcFace was trained on MS1M with plain shuffling.
This is the selling point over pair-based losses. Note that sampling starts to matter in the
long-tailed re-ID regime and defer that to the companion notebook. **Do not imply ArcFace
needs a class-balanced sampler in general.**

### Required baselines

The notebook is unfalsifiable without these. Compute all three and report side by side:

1. **Frozen ImageNet features + cosine kNN.** No metric training at all. Strong on CUB.
2. **Plain softmax cross-entropy** on the 100 training classes, penultimate layer as
   embedding.
3. **ArcFace.**

Frame the notebook as "measure, don't assume". Cite Musgrave et al. (2020), *A Metric Learning
Reality Check* — claimed gains largely vanish under fair comparison. **Do not write prose that
assumes ArcFace wins.** It may win narrowly or not at all. If it doesn't, that is the lesson,
and the narrative must be written after the numbers exist, not before.

### Metrics

Recall@1, Recall@5, Recall@10 (excluding the query itself) and mAP@R. Implement both; mAP@R
is the Musgrave recommendation and the more honest of the two.

### Compute

Precompute backbone features once and cache as `.npy`, so the head trains in seconds on CPU.
Note in the notebook that frozen features mute the margin's effect, and that fine-tuning the
last block is the compromise used in the companion notebook.

### CLIP zero-shot — the payoff section

Inference only, no training.

**Framing:** ArcFace learns class proxies from labelled data. CLIP's text encoder *generates*
proxies from language. Zero-shot classification is nearest-proxy retrieval in a shared
image–text space. Make the parallel to `ArcFaceHead`'s weight matrix explicit.

- KerasHub ships CLIP presets (`clip_vit_base_patch16`, `clip_vit_base_patch32`). Use these
  so we stay in Keras 3 / JAX. **Verify the exact task-class API on the current `keras-hub`
  before writing cells** — check whether a ready zero-shot task class exists, or whether the
  image and text encoders plus `logit_scale` must be composed manually. Add `keras-hub` to
  the environment spec.
- Run on the 100 held-out CUB test classes, so the comparison to the supervised metric model
  is like-for-like.
- **Prompt engineering:** compare bare class names against `"a photo of a {}, a type of
  bird"`. The gap is several points and is the point of the exercise.
- **Expected result and honest framing:** zero-shot accuracy on fine-grained species will be
  mediocre, and the supervised metric model trained on 100 classes should beat it in domain.
  That is the correct lesson — fine-grained taxonomy is near CLIP's limit. Do not oversell.
- **Contamination caveat:** CUB images plausibly sit in CLIP's training data. One sentence.
- **BioCLIP:** mention in a markdown cell as the fine-grained taxonomic alternative, noting it
  has no KerasHub preset so we are not running it. Do **not** propose WildCLIP for CUB — it is
  camera-trap and attribute-vocabulary oriented (Snapshot Serengeti) and may not transfer.
- **Cache the CLIP image and text embeddings as `.npy` and commit them.** The companion
  notebook reuses the image encoder on turtle images and should not have to recompute.

---

## Deliverables

- [ ] `sessions/metric_learning.ipynb`, executed, with outputs
- [ ] Standalone GPU training script(s) for the softmax and ArcFace models
- [ ] `.keras` checkpoints committed for both trained models
- [ ] Precomputed arrays committed: CUB backbone features, CLIP image/text embeddings
- [ ] `index.ipynb` updated
- [ ] Environment spec updated (`keras-hub`)
- [ ] Every trap in the traps section verified on Keras 3.15.1 and documented

## Acceptance criteria

- Runs top-to-bottom on CPU in a few minutes with training cells commented out.
- No PyTorch dependency.
- All three baselines computed and reported side by side; narrative does not assume ArcFace
  wins.
- `ArcFaceHead` and the margin loss use `keras.ops` only and run under the JAX backend.
- Opening cell states the delta from `functional_keras.ipynb`.

## Non-goals

- Training or fine-tuning CLIP or BioCLIP.
- Triplet/contrastive loss implementation (conceptual only; the Siamese notebook has the
  pair-based treatment).
- Few-shot episodic evaluation — moved to the companion re-ID notebook, where the long-tailed
  identity distribution makes it more honest.
- Any re-identification content.

Question now answered: Which backbone does `transfer.ipynb` currently use? 
Answer: EfficientNetV2S

Replaces #7

**Comment by yoavram** [comment]
See branch metric-learning

**Comment by yoavram** [comment]
## Status: complete on branch `metric-learning`, with one loose end

Branch `metric-learning` (off `DL2026`), 7 commits, `29b4aef` / `b655231` at the tip. The notebook runs top to bottom on CPU in **4.4 minutes** with nothing commented out, 82 cells, zero errors.

### Deviations from this issue, all deliberate

- **No ArcFace.** The angular margin measured as nothing in every regime tested — including, once the comparison was controlled for scale, on frozen EfficientNetV2S features. The notebook builds the normalized rescaled proxy layer (`CosineHead`) and stops there; the margin is Exercise 3 and moves properly to #12. See the [comment on #12](https://github.com/yoavram/DataSciPy/issues/12#issuecomment-5749902221) for what that changes there. This drops the "ArcFace" baseline and the `keras.ops` margin-loss acceptance criterion.
- **Checkpoints and precomputed arrays are not committed.** `.gitignore` and `CLAUDE.md` forbid committing weights and arrays. They are produced by `scripts/metric_learning_features.py`; `download_data.py` gained the CLIP preset.
- **`keras-hub` is not in `requirements.txt`.** pip cannot express per-package `--no-deps`, and keras-hub declares `tensorflow-text`. Its real dependencies are listed; the `--no-deps` install is a README step and a `TROUBLESHOOTING.md` entry.
- **Prompt templating does not help on CUB**, contrary to this issue's expectation. Selecting honestly on held-out species picks a template that then scores 51.06% against the canonical template's 51.11%, with Spearman +0.20 between the two rankings. That negative result is now one of the strongest sections.

### Open item — one unverified docstring

`solutions/metric_learning_clip_finetune.py` (the worked solution to Exercise 7) was verified end to end on an A4000 **before** the margin was dropped, producing:

```
softmax        18 epochs   R@1 71.03%   mAP@R 31.33%
cosine head     2 epochs   R@1 73.02%   mAP@R 34.90%
```

The script now searches scales only. Those two rows are not expected to move — the selected configuration was already the best margin-free one — but they are **unconfirmed under the trimmed search**, and the docstring says so.

**To close this:** re-run `KERAS_BACKEND=jax python solutions/metric_learning_clip_finetune.py` (about 25 minutes on an A4000, three search runs plus two refits) and replace the `VERIFICATION PENDING` block in its docstring with the confirmed output. If the numbers differ materially, the fine-tune paragraph in the notebook's discussion needs revisiting too.

A run was in flight when this session ended; its log, if the machine still has it, is at
`/tmp/claude-1004/-home-jupyter-yoavram-Work-Teaching-DataSciPy/6c4b383a-9de3-450b-9ff4-fec04b15b818/scratchpad/clip_ft_final.log`. Do not trust it without checking it ran to completion — simplest is to just re-run.

### Also still open

The notebook's opening cell compares itself to a Siamese treatment in `sessions/functional_keras.ipynb` that currently exists only on the unmerged `siamese` branch. That forward reference is dead until `siamese` lands on `DL2026`.

**Comment by yoavram** [comment]
### Open item closed — `6d2f5e2`

The re-run finished before the session ended. The numbers did not move, as expected: the configuration the trimmed search selects was already the best margin-free one.

```
softmax      softmax   18 epochs   R@1 71.02%   mAP@R 31.33%
cosine head  s=8        2 epochs   R@1 73.02%   mAP@R 34.90%
```

against 69.0% / 29.6% and 70.2% / 31.9% for the same two heads on frozen CLIP features in the notebook. `VERIFICATION PENDING` is replaced with the confirmed output, and no notebook prose needed revisiting — the fine-tune paragraph in the discussion is unaffected.

Two details from the run were worth adding to the docstring:

- **The selected scale moves between regimes** — s=4 on frozen CLIP features, s=8 once the backbone is unfrozen. That is precisely why the search is re-run in-regime rather than carried over, and it is the same lesson that cost the notebook a rewrite earlier.
- **The budgets collapse.** The softmax head wants 18 epochs; the cosine head peaks after **two** and is past its best by the third.

Nothing outstanding on this issue now except the `functional_keras.ipynb` forward reference, which is blocked on the `siamese` branch landing on `DL2026`.

**Comment by yoavram** [comment]
Done on `DL2026`.

**Added by [`6534255`](https://github.com/yoavram/DataSciPy/commit/6534255) — "Add a session on open-set metric learning and CLIP zero-shot (#11)".**

`sessions/metric_learning.ipynb`: CUB-200 split by class, frozen-feature baseline, the softmax-then-discard-the-classifier result, Recall@K and mAP@R, the stopping-epoch argument, and CLIP zero-shot with prompts measured on held-out species.

Refined across nine follow-ups, of which these changed what the session claims rather than how it reads:

- [`01f6843`](https://github.com/yoavram/DataSciPy/commit/01f6843) — controlled the margin comparison
- [`29b4aef`](https://github.com/yoavram/DataSciPy/commit/29b4aef) — dropped the angular margin, kept the normalized cosine head
- [`3e1630c`](https://github.com/yoavram/DataSciPy/commit/3e1630c) — fixed a search/refit schedule mismatch
- [`3d44a8d`](https://github.com/yoavram/DataSciPy/commit/3d44a8d) — raised `MAX_EPOCHS`; the softmax arm on CLIP had been selecting its own cap
- [`21d5fd9`](https://github.com/yoavram/DataSciPy/commit/21d5fd9) — cut the session back to its spine: one idea, measured once
- [`45d2bd0`](https://github.com/yoavram/DataSciPy/commit/45d2bd0) — three seeds on the LP-FT row, and raised the cap it was hiding
- [`299ee98`](https://github.com/yoavram/DataSciPy/commit/299ee98) — took the cosine head out of the fine-tune script too

The cosine head and angular margin were deliberately cut from this session and are now built from scratch in `sessions/reid.ipynb` (#12), which this notebook forward-references.
